cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java

```diff
@@ -103,35 +105,32 @@ protected long getInitialDelayMillis() {
 
   @Override
   protected void startUp() throws Exception {
-    runThread = Thread.currentThread();
+    // No-op
   }
 
   @Override
   protected final void run() {
     long sleepMillis = getInitialDelayMillis();
     while (isRunning()) {
-      try {
-        TimeUnit.MILLISECONDS.sleep(sleepMillis);
-        long startTime = System.currentTimeMillis();
-        publishMetrics(startTime);
-        sleepMillis = Math.max(0, publishIntervalInMillis - (System.currentTimeMillis() - startTime));
-      } catch (InterruptedException e) {
-        // Expected when stop is called.
+      if (Uninterruptibles.awaitUninterruptibly(shutdownLatch, sleepMillis, TimeUnit.MILLISECONDS)) {
+        // If the shutdown latch is triggered, it means triggerShutdown has been called, hence break the loop
         break;
       }
+
+      long startTime = System.currentTimeMillis();
+      publishMetrics(startTime);
```

Comment:
this means if the publish interval is 5 seconds, it may take up to 5 seconds to shutdown. Is that good? Or could we sleep in shorter intervals repeatedly, to check sooner?

Reply:
We use `CountDownLatch` to signal the shutdown. It will return immediately. I am assuming the call the `publishMetrics` is relatively fast (basically publishing to TMS), which doesn't have sleep involved.